### DISCUSSION TITLE
spu与heu的两个问题

### DISCUSSION THREAD

**Discussion by deadlywing** [secretflow/secretflow#42] [Q&A] [answered]
我在看群里今天发的文章，有以下几个问题:

# HEU支持的算法
**Q1** 文章里HEU前加了`(半同态)`，是现在只支持半同态的加密算法么？后续有没有支持比如FBV,CKKS这些全同态算法呢？
<img width="747" alt="image" src="https://user-images.githubusercontent.com/25818621/179133718-e4444932-4457-461f-96dc-d2290b9f688a.png">

# HEU和SPU之间的转换协议

**Q2：** `heu`和`spu`之间的转换协议的考虑；比如`spu`到`heu`， 因为我的第一想法肯定是要么直接对share加密，要么就是reconstruction以后加密。后者肯定不合理，前者的话似乎对arithemetic share是ok的？

**Q.3：** 对`heu`到`spu`，我能想到的一个比较类似的做法是[sshelr](https://arxiv.org/pdf/2008.08753.pdf)这篇paper里的`Protocol2`，就是从半同态密文到share的转换，操作也是解密和同态减法；但是`spu`到`heu`我想不太到是什么思路，能麻烦大概解释一下或者给一下参考的paper么？

<img width="751" alt="image" src="https://user-images.githubusercontent.com/25818621/179133495-854cae26-58e4-452a-b373-eb0388783589.png">

**Comment by rivertalk** [comment, marked as answer]
Hi @deadlywing 

非常好的问题！我尝试着解答一下

Q1: 目前是只支持半同态算法，全同态算法在规划中，但暂时还没有具体的发布计划
Q2: 是的，直接在Share上加密和转化，中间不会reconstruction，对Arithmetic Share是OK的
Q3: 是的，算法就是上述sshelr中的Protocol2。从spu到heu的过程和Protocol2类似，A将分片加密发送给B，B在同态空间求和，这样A拥有私钥，B拥有加密数据，是一个Protocol2的反过程

**Reply by deadlywing** [reply]
感谢解答。我还有两个小问题，
1. 那spu到heu的转换就不对称了呀，怎么决定谁是秘钥拥有的那方呢，这个是否是可配置的？（虽然理论上我感觉谁有秘钥谁有密文没区别，但是在实际应用的时候，对两个算力不对等的client而言，这个选择有时候还是能影响整体的效率的）
2. 现在heu里支持的半同态算法是Paillier么？还是dgk什么的？

**Reply by rivertalk** [reply]
谢谢提问，都是非常专业的问题！

1. 是的，spu到heu转换之后是不对称了，算法实现者需要手动指定密钥拥有方，这个是可配置的。的确，谁拥有密钥谁拥有密文对安全性的确是没有区别的，但是会影响到效率
2. heu 里支持的半同态算法是 Paillier，具体见 https://github.com/secretflow/heu/blob/beta/heu/library/phe/schema.h

**Reply by deadlywing** [reply]
非常感谢您的耐心解答！
